library/queue.lisp

```diff
@@ -0,0 +1,258 @@
+  (define (copy q)
+    "Return a new queue containing the same elements as `q`."
+    (lisp (Queue :a) (q)
+      (cl:multiple-value-bind (elements last)
+          ;; Copy the list, holding on to a reference to the last cons
+          (cl:loop
+             :with head := (cl:cons (cl:car (queue-internal-elements q)) cl:nil)
+             :with tail := head
+             :for e :in (cl:cdr (queue-internal-elements q))
+             :for new-cons := (cl:cons e nil)
+             :do (cl:setf (cl:cdr tail) new-cons
+                          tail new-cons)
+             :finally (cl:return (cl:values head tail)))
+        (make-queue-internal
+         :elements elements
+         :last last
+         :length (queue-internal-length q)))))
+
+  (declare clear! (Queue :a -> Unit))
+  (define (clear! q)
+    "Clear all elements from `q`."
+    (lisp Unit (q)
+      (cl:setf (queue-internal-elements q) cl:nil
+               (queue-internal-last q) cl:nil
+               (queue-internal-length q) 0)
+      Unit))
+
+  (declare push! (:a -> Queue :a -> Unit))
+  (define (push! item q)
+    "Push `item` onto the end of `q`."
+    (lisp Unit (item q)
+      (cl:let ((last (cl:cons item cl:nil)))
+        (cl:if (cl:null (queue-internal-elements q))
+               ;; Set up the queue with the first element. Note that the same
+               ;; reference to the singleton list is shared by both
+               ;; QUEUE-INTERNAL-ELEMENTS and QUEUE-INTERNAL-LAST.
+               (cl:setf (queue-internal-elements q) last
+                        (queue-internal-last q) last)
+
+               ;; We can now append elements to QUEUE-INTERNAL-ELEMENTS simply by
+               ;; modifying QUEUE-INTERNAL-LAST, whose reference is shared by
+               ;; QUEUE-INTERNAL-ELEMENTS,
+               ;;
+               ;; We do this instead of a single SETF for type safety of
+               ;; QUEUE-INTERNAL-LAST.
+               (cl:let ((old (queue-internal-last q)))
+                 (cl:setf (queue-internal-last q) last
+                          (cl:cdr old) last)))
+
+        (cl:incf (queue-internal-length q)))
+
+      Unit))
+
+  (declare pop! (Queue :a -> Optional :a))
+  (define (pop! q)
+    "Remove and return the first item of `q`."
+    (if (empty? q)
+        None
+        (Some (pop-unsafe! q))))
+
+  (declare pop-unsafe! (Queue :a -> :a))
+  (define (pop-unsafe! q)
+    "Remove and return the first item of `q` without checking if the queue is empty."
+    (lisp :a (q)
+      (cl:decf (queue-internal-length q))
+      (cl:pop (queue-internal-elements q))))
+
+  (declare peek (Queue :a -> Optional :a))
+  (define (peek q)
+    "Peek at the first item of `q`."
+    (if (empty? q)
+        None
+        (Some (peek-unsafe q))))
+
+  (declare peek-unsafe (Queue :a -> :a))
+  (define (peek-unsafe q)
+    "Peek at the first item of `q` without checking if the queue is empty."
+    (lisp :a (q)
+      (cl:car (queue-internal-elements q))))
```

Comment:
this could be a type error, which i guess is par for the course for UNSAFE

Reply:
It seems better for it to error rather than return invalid data. Will fix